Add support for Ethereum EIP 1559

FrederikBolding opened the pull request:
- Updated Trezor Common to include new EIP 1559 message
- Add new method for signing EIP 1559 transactions

FrederikBolding committed:
Update Trezor Common

FrederikBolding committed:
Initial JS implementation

FrederikBolding committed:
Fix wrong arguments

FrederikBolding commented:
Heya! Hoping to start a conversation about how to implement https://github.com/trezor/trezor-firmware/pull/1653 in the JS library.

FrederikBolding commented:
@szymonlesisz Sorry for the ping, but looking to start the conversation about implementing this.

I haven't figured out how to test with the new firmware build, but I think the code here should work with that.

Curious to hear your thoughts on whether you would rather use the same function for both the old and new transaction types.

I can also add more test cases once we get some progress on this.

matejkriz commented:
Hi @FrederikBolding , thanks for the draft! 

We were thinking more of keeping only `EthereumSignTx` method, the difference is really small and would be nice to handle the selection of protobuf message here inside trezor-connect. 

What mechanism are you planning to use in your app for detection if firmware of connected device supports the new message? For Suite we would like to define new unavailable capability here https://github.com/trezor/connect/blob/develop/src/data/config.json#L106 . So the condition for protobuf message selection could look something like `!device.unavailableCapabilities.eip1559 && !!max_priority_fee ? EthereumSignTxEIP1559 : EthereumSignTx`.

I could help you to run the new firmware in emulator, if you are interested. Starting point is https://github.com/trezor/trezor-firmware/blob/master/docs/core/build/emulator.md

FrederikBolding commented:
> Hi @FrederikBolding , thanks for the draft!
> 
> We were thinking more of keeping only `EthereumSignTx` method, the difference is really small and would be nice to handle the selection of protobuf message here inside trezor-connect.

Sounds good to me, this is my preferred solution as well. I will make appropriate changes for this!

> What mechanism are you planning to use in your app for detection if firmware of connected device supports the new message? For Suite we would like to define new unavailable capability here https://github.com/trezor/connect/blob/develop/src/data/config.json#L106 . So the condition for protobuf message selection could look something like `!device.unavailableCapabilities.eip1559 && !!max_priority_fee ? EthereumSignTxEIP1559 : EthereumSignTx`.

We haven't thought to much about that to be honest, but if you have a good way of doing that I don't have problem with it! What would be the format of adding that capability?

> I could help you to run the new firmware in emulator, if you are interested. Starting point is https://github.com/trezor/trezor-firmware/blob/master/docs/core/build/emulator.md

I have the new firmware running in an emulator from when I wrote the firmware PR. But I am unsure of how this repos tests with it 😄

FrederikBolding committed:
Simplify implementation to use one method

FrederikBolding committed:
Simplify further

FrederikBolding committed:
Dont do chain id calculation for EIP 1559

FrederikBolding commented:
@matejkriz I have simplified to use the same method for both calls. Still awaiting some instructions on tests as well as `unavailableCapabilities`.

matejkriz commented:
@FrederikBolding regarding the testing, when you [run this repo locally (`yarn run dev`)](https://wiki.trezor.io/Developers_guide:Running_Trezor_Connect_on_localhost), you need to visit [https://localhost:8088/](https://localhost:8088/) first in your browser to confirm the certificate and then you could use it in your app or use connect-explorer: [https://trezor.github.io/connect-explorer/?trezor-connect-src=https://localhost:8088/#/method/ethereumSignTransaction](https://trezor.github.io/connect-explorer/?trezor-connect-src=https://localhost:8088/#/method/ethereumSignTransaction) (it needs page reload on any change in trezor-connect)

matejkriz commented:
Format for the new unavailableCapability in https://github.com/trezor/connect/blob/develop/src/data/config.json#L146 could be 
```
,
{
  "excludedMethods": ["eip1559"],
  "min": ["0", "2.4.2"],
  "comment": ["new eth transaction pricing mechanism (EIP1559) since 2.4.2"]
}
```

FrederikBolding commented:
> @FrederikBolding regarding the testing, when you [run this repo locally (`yarn run dev`)](https://wiki.trezor.io/Developers_guide:Running_Trezor_Connect_on_localhost), you need to visit https://localhost:8088/ first in your browser to confirm the certificate and then you could use it in your app or use connect-explorer: https://trezor.github.io/connect-explorer/?trezor-connect-src=https://localhost:8088/#/method/ethereumSignTransaction (it needs page reload on any change in trezor-connect)

I've gotten this far, but how do I force Trezor Connect to talk to the emulator instead of a normal device and how do I run the tests on the emulator?

FrederikBolding committed:
Add unavailableCapabilities

szymonlesisz commented:
Hi @FrederikBolding ,

First of all: __WOW__  big thanks for the integration!
I will have some small nit-picks later but it looks really nice at first glance :)

you can run tests with emulator `./tests/run.sh -i ethereumSignTransaction` (see [tests docs](https://github.com/trezor/connect/blob/develop/tests/README.md))

or if you wish to work with the emulator during integration im recommending https://github.com/trezor/trezor-user-env docker image - it contains multiple firmwares including `2.4.2` (2-master)

FrederikBolding committed:
Update test case

FrederikBolding committed:
Add access lists

FrederikBolding committed:
Small access list fixes

FrederikBolding committed:
Fix access_list fixture

szymonlesisz commented on tests/__fixtures__/ethereumSignTransaction.js:
```diff
@@ -358,5 +358,54 @@ export default {
                 v: '0x2a',
             },
         },
+
```
lets keep it compatible with fixtures from  https://github.com/trezor/trezor-firmware/blob/master/tests/device_tests/test_msg_ethereum_signtx_eip1559.py

FrederikBolding replied:
These tests use another mnemonic, so I can't use the same ones. I am just generating new test cases with the new mnemonic. Is that fine?

szymonlesisz replied:
ah, i get it, the are using `mnemonic_all`. so either you can create another file next to `__fixtures__/ethereumSignTransaction.js` let say `ethereumSignTransactionEip1559.js` or you can set mnemonic for single case like:

```
{
            description: 'no data eip 1559',
            mnemonic: 'mnemonic_all`,
            params: ...
```

FrederikBolding replied:
Okay, I can port the tests over if you prefer that 😄

FrederikBolding replied:
Done!

szymonlesisz commented on src/js/types/networks/ethereum.js:
```diff
@@ -36,11 +36,30 @@ export type EthereumTransaction = {
+export type EthereumAccessList = {
+    address: string;
+    storageKeys: string[];
+};
```
please add this new types to typescript definitions as well. we provide both, crazy - i know :)

https://github.com/trezor/connect/blob/develop/src/ts/types/networks/ethereum.d.ts

FrederikBolding replied:
Done!

FrederikBolding committed:
Add more tests

FrederikBolding committed:
Add TS types

FrederikBolding committed:
Fix stripping on access list

FrederikBolding committed:
Update fixtures

szymonlesisz commented on src/data/messages/messages.json:
```diff
@@ -10599,6 +10595,117 @@
             "options": {},
             "oneofs": {}
+        },
+        {
```
i'm wondering why it was shifted like that?
did you generate protobuf using `./scripts/protobuf-build.sh local`?

i did this on your branch and it's shifting it back where it was

FrederikBolding replied:
I have no idea. I just re-ran that command and it changed a few things, but not this.

FrederikBolding replied:
If you want, you can push it to my branch 😄

FrederikBolding committed:
Update fixture

FrederikBolding committed:
Run protobuf gen again

FrederikBolding committed:
Fix a few small mistakes in fixtures

FrederikBolding committed:
Fix linting

FrederikBolding committed:
Update fixture

FrederikBolding committed:
Update fixture

FrederikBolding commented:
What I have now works and passes tests with fixtures copy-pasted from the test cases I created in the firmware. Good to start an actual review. 

Let me know if you have any questions 😄

szymonlesisz commented:
@FrederikBolding ,

Sorry for taking so long but i've found a problem with `unavailableCapabilities` so i needed to fix it  [here](https://github.com/trezor/connect/pull/875) before merging this.
TLDR: `eip1559` unavailableCapability added here was not shown properly


I've rebased your branch on top of my PR with the fix, squashed all your commits into one and added some fixups from me:

https://github.com/trezor/connect/pull/874/commits


in f94aecf1cbac09afbdc8257d25b01e04f6fa45d6
i've changed field name in config.json (related to PR mentioned above) + added few unit tests

in 1dcdeaad816f553fcc6e03e1f9b70919cc6be81b
i've changed flowtype/typescript definitions a litte bit + added few types tests

in e216b0e3fec3a9c720f7fe57b9e355e1e75b9565
i've updated the code to validate firmware range before running method (to prevent signing eip1559 with T1 - which is not supported) + minor changes to make flowtype happy 

in 1b56b9ebf36a62723bf5fa1c6219c2e638b582c2
i've moved eip1559 fixtures into standalone file + added 2 missing cases from firmware repo (erc20 transfer)


Now all tests are passing
Let me know if you are ok with my fixups so i can squash them

FrederikBolding commented:
@szymonlesisz 

> @FrederikBolding ,
> 
> Sorry for taking so long but i've found a problem with `unavailableCapabilities` so i needed to fix it [here](https://github.com/trezor/connect/pull/875) before merging this.
> TLDR: `eip1559` unavailableCapability added here was not shown properly
> 
> I've rebased your branch on top of my PR with the fix, squashed all your commits into one and added some fixups from me:
> 
> https://github.com/trezor/connect/pull/874/commits

No worries about the wait!

> in [e216b0e](https://github.com/trezor/connect/commit/e216b0e3fec3a9c720f7fe57b9e355e1e75b9565)
> i've updated the code to validate firmware range before running method (to prevent signing eip1559 with T1 - which is not supported) + minor changes to make flowtype happy

It seems you removed `this.device.unavailableCapabilities.eip1559`, how is it then detected if a user tries to send a type 2 tx without having the proper firmware? I might be misunderstanding how the `firmwareRange` code works?

> Now all tests are passing
> Let me know if you are ok with my fixups so i can squash them

Other than the small question above, everything else looks good to me. You can push your squashed commits to my branch if you want. I don't know what you prefer.

Thanks in advance!!

szymonlesisz commented:
> I might be misunderstanding how the `firmwareRange` code works?



validation of capabilities in `EthereumSignTranaction.run` is done "too late", from this place you can either return success or failure, there is no "interaction with user" to tell him what is going on.

`firmwareRange` check is [used before](https://github.com/trezor/connect/blob/develop/src/js/core/Core.js#L467) `EthereumSignTranaction.run` and user will be noticed in popup about required update

![Screenshot from 2021-08-18 15-15-34](https://user-images.githubusercontent.com/3435913/129905754-f4db9712-1031-4e9e-9f47-4f7dedda5d4f.png)

checking for `i can use it` + `i have eip1559 params`  gives wrong result when `i CANT use it` + `but i have eip1559` params example: running ethereumSignTransaction with eip1559 params on 2.4.1 where support is disabled by config.json will fall to the legacy signing block and will use `eip1559 params` in `helper.ethereumSignTx`

FrederikBolding commented:
> > I might be misunderstanding how the `firmwareRange` code works?
> 
> 
> 
> validation of capabilities in `EthereumSignTranaction.run` is done "too late", from this place you can either return success or failure, there is no "interaction with user" to tell him what is going on.
> 
> `firmwareRange` check is [used before](https://github.com/trezor/connect/blob/develop/src/js/core/Core.js#L467) `EthereumSignTranaction.run` and user will be noticed in popup about required update
> 
> ![Screenshot from 2021-08-18 15-15-34](https://user-images.githubusercontent.com/3435913/129905754-f4db9712-1031-4e9e-9f47-4f7dedda5d4f.png)
> 
> checking for `i can use it` + `i have eip1559 params`  gives wrong result when `i CANT use it` + `but i have eip1559` params example: running ethereumSignTransaction with eip1559 params on 2.4.1 where support is disabled by config.json will fall to the legacy signing block and will use `eip1559 params` in `helper.ethereumSignTx`
> 
> 
> 

Sounds good!

FrederikBolding commented:
Closing in favor of #874

puruoni commented:
Hi, thanks for the EIP-1559 support!
I'd like to use it with Trezor One, but it's not possible?

> ⚠️ Supported only by Trezor T with Firmware 2.4.2 or higher!

FrederikBolding commented:
> Hi, thanks for the EIP-1559 support!
> I'd like to use it with Trezor One, but it's not possible?
> 
> > ⚠️ Supported only by Trezor T with Firmware 2.4.2 or higher!

Unfortunately it is not supported!

FrederikBolding commented:
@szymonlesisz Sorry for the ping, but not sure where else to contact. Is this supposed to be releasing today? I saw the library was updated, but using the latest version throws `Parameter \"gasPrice\" is missing.`, so I assume not everything is updated. How/when is the update coming out?

shroom-dev commented:
Hi, will EIP-1559 support be added to Trezor One at some point soon?

greenlucid commented:
Is it never going to be supported for Trezor One?
That would be so lame. You know there's an offer from Eth core multisig for this, right?

CharlesSOo commented:
No comment?

liszper commented:
@FrederikBolding I upgraded to 2.4.2, latest MetaMask too, but my Trezor T still doesn't support direct "write contract" on etherscan.io, getting "Invalid transaction params: params specify an EIP-1559 transaction but the current network does not support EIP-1559" with the latest Metamask version or very high gas costs (indicating there is an error) with older versions of Metamask.

I've spent months trying to solve this with dozens of MetaMask and Trezor firmware versions and now It's becoming a significant problem for me. Could you help me please?

FrederikBolding commented:
> @FrederikBolding I upgraded to 2.4.2, latest MetaMask too, but my Trezor T still doesn't support direct "write contract" on etherscan.io, getting "Invalid transaction params: params specify an EIP-1559 transaction but the current network does not support EIP-1559" with the latest Metamask version or very high gas costs (indicating there is an error) with older versions of Metamask.
> 
> I've spent months trying to solve this with dozens of MetaMask and Trezor firmware versions and now It's becoming a significant problem for me. Could you help me please?

I'm not quite sure whether MetaMask has rolled out 1559 support for Trezor at this point. 

But if it is not working for you, you can use https://app.mycrypto.com/interact-with-contracts to interact with contracts using Trezor Model T. If you have any issues feel free to reach out on our Discord or to our support team!

guff-se commented:
Wow. Just bought a Trezor One for the specific purpose of managing Ethereum contracts. Am I understanding it correctly that there is no plan for EIP-1559 support for the One? So I can basically trash it and buy something else? :'(
